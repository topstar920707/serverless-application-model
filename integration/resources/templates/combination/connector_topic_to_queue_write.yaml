Resources:
  LambdaRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Action: sts:AssumeRole
            Principal:
              Service: lambda.amazonaws.com
      ManagedPolicyArns:
        - !Sub arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole

  TriggerFunction:
    Type: AWS::Lambda::Function
    Properties:
      Role: !GetAtt LambdaRole.Arn
      Runtime: nodejs14.x
      Handler: index.handler
      Code:
        ZipFile: |
          const AWS = require('aws-sdk');
          var params = {
            Message: 'Test Message',
            TopicArn: process.env.TOPIC_ARN,
          };
          var sqs_params = {
            QueueUrl: process.env.QUEUE_NAME,
          };
          exports.handler = async (event) => {
            var sns = new AWS.SNS({apiVersion: '2012-11-05'});
            await sns.publish(params).promise();

            var sqs = new AWS.SQS({apiVersion: '2012-11-05'});
            var data = await sqs.receiveMessage(sqs_params).promise();
            if (data.Messages.length == 0) {
              throw 'No messages in the queue!';
            }
          };
      Environment:
        Variables:
          TOPIC_ARN: !Ref Topic
          QUEUE_NAME: !Ref Queue

  Topic:
    Type: AWS::SNS::Topic
    Properties:
      Subscription:
        - Endpoint:
            Fn::GetAtt:
              - "Queue"
              - "Arn"
          Protocol: "sqs"

  Queue:
    Type: AWS::SQS::Queue

  Connector:
    Type: AWS::Serverless::Connector
    Properties:
      Source:
        Id: TriggerFunction
      Destination: 
        Id: Topic
      Permissions:
        - Write

  Connector2:
    Type: AWS::Serverless::Connector
    Properties:
      Source:
        Id: Topic
      Destination: 
        Id: Queue
      Permissions:
        - Write

  Connector3:
    Type: AWS::Serverless::Connector
    Properties:
      Source:
        Id: TriggerFunction
      Destination: 
        Id: Queue
      Permissions:
        - Read