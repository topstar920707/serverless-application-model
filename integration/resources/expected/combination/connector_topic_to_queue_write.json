[
  {
    "LogicalResourceId": "LambdaRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "TriggerFunction",
    "ResourceType": "AWS::Lambda::Function"
  },
  {
    "LogicalResourceId": "Queue",
    "ResourceType": "AWS::SQS::Queue"
  },
  {
    "LogicalResourceId": "Topic",
    "ResourceType": "AWS::SNS::Topic"
  },
  {
    "LogicalResourceId": "ConnectorPolicy",
    "ResourceType": "AWS::IAM::ManagedPolicy"
  },
  {
    "LogicalResourceId": "Connector2QueuePolicy",
    "ResourceType": "AWS::SQS::QueuePolicy"
  },
  {
    "LogicalResourceId": "Connector3Policy",
    "ResourceType": "AWS::IAM::ManagedPolicy"
  }
]